src/mcp23017.cpp

```diff
@@ -1,41 +1,114 @@
 #include "mcp23017.h"
-#include <errno.h>
-
 
 // TODO (optional): Define macros for useful register below:
-
+#define IODIRA 0x00
+#define IODIRB 0x01
+#define GPIOA 0x12
+#define GPIOB 0x13
 
 // TODO: Initialize i2cBus member
-Mcp23017::Mcp23017(int addr, I2C* i2cBus) {
-
+Mcp23017::Mcp23017(int addr) {
+    this->addr = addr;
 }
 
 uint8_t Mcp23017::get_dir(int pin) {
-    return 0;
+    Wire.beginTransmission(this->addr);
+    uint8_t dataReg = IODIRA;
+    Wire.write(dataReg);
+    Wire.endTransmission();
+
+    Wire.requestFrom(this->addr, 1);
+
+    int data = Wire.read();
+
+    int p = 1 & (data >> pin);
+    return p;
 }
 
 
 // TODO: Read from state register
 uint8_t Mcp23017::get_state(int pin) {
-    return 0;
+    Wire.beginTransmission(this->addr);
+    uint8_t dataReg = GPIOA;
+    Wire.write(dataReg);
+    Wire.endTransmission();
+
+    Wire.requestFrom(this->addr, 1);
+
+    int data = Wire.read();
+
+    int p = 1 & (data >> pin);
+    return p;
 }
 
 // TODO: Write to directions register
 int Mcp23017::set_dir(int pin, uint8_t dir) {
-    return 0;
+    int byte = get_dir(pin);
```

Comment:
`get_dir()` doesn't return the entire byte that's in the IODIRA register, so this implementation doesn't correctly modify only the specific pin we want to modify.